key.go

```diff
@@ -0,0 +1,777 @@
+		expectedAlg, err := k.deriveAlgorithm()
+		if err != nil {
+			return err
+		}
+
+		if k.Algorithm != expectedAlg {
+			return fmt.Errorf(
+				"found algorithm %q (expected %q)",
+				k.Algorithm.String(),
+				expectedAlg.String(),
+			)
+		}
+	}
+
+	return nil
+}
+
+// MarshalCBOR encodes Key into a COSE_Key object.
+func (k Key) MarshalCBOR() ([]byte, error) {
+	var err error
+
+	if k.KeyType == KeyTypeSymmetric {
+		if k.RawKeyParam, err = encMode.Marshal(k.K); err != nil {
+			return nil, err
+		}
+	} else if k.KeyType == KeyTypeEc2 || k.KeyType == KeyTypeOkp {
+		if k.RawKeyParam, err = encMode.Marshal(k.Curve); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("invalid key type: %q", k.KeyType.String())
+	}
+
+	return encMode.Marshal(k.keyStruct)
+}
+
+// UnmarshalCBOR decodes a COSE_Key object into Key.
+func (k *Key) UnmarshalCBOR(data []byte) error {
+	if err := decMode.Unmarshal(data, &k.keyStruct); err != nil {
+		return err
+	}
+
+	switch k.KeyType {
+	case KeyTypeEc2:
+		if k.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for EC2 key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeOkp:
+		if k.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for OKP key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeSymmetric:
+		if k.RawKeyParam == nil {
+			return errors.New("missing K parameter (required for Symmetric key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.K); err != nil {
+			return err
+		}
+	default:
+		// this should not be reachable as KeyType.UnmarshalCBOR would
+		// result in an error during decMode.Unmarshal() above, if the
+		// value in the data doesn't correspond to one of the above
+		// types.
+		return fmt.Errorf("unexpected key type %q", k.KeyType.String())
+	}
+
+	return k.Validate()
+}
+
+// GetPublic returns a crypto.PublicKey generated using Key's parameters.
+func (k *Key) GetPublic() (crypto.PublicKey, error) {
+	alg, err := k.deriveAlgorithm()
+	if err != nil {
+		return nil, err
+	}
+
+	switch alg {
+	case AlgorithmES256, AlgorithmES384, AlgorithmES512:
+		var curve elliptic.Curve
+
+		switch alg {
+		case AlgorithmES256:
+			curve = elliptic.P256()
+		case AlgorithmES384:
+			curve = elliptic.P384()
+		case AlgorithmES512:
+			curve = elliptic.P521()
+		}
+
+		pub := &ecdsa.PublicKey{Curve: curve, X: new(big.Int), Y: new(big.Int)}
+		pub.X.SetBytes(k.X)
+		pub.Y.SetBytes(k.Y)
+
+		return pub, nil
+	case AlgorithmEd25519:
+		return ed25519.PublicKey(k.X), nil
+	default:
+		return nil, ErrAlgorithmNotSupported
+	}
+}
+
+// GetPrivate returns a crypto.PrivateKey generated using Key's parameters.
+func (k *Key) GetPrivate() (crypto.PrivateKey, error) {
+	alg, err := k.deriveAlgorithm()
+	if err != nil {
+		return nil, err
+	}
+
+	switch alg {
```

Comment:
And here as well!

Reply:
Note that this is the reverse of the other case (here it's curve-to-algo), so would not be re-used. Given that it's used in only one place, already factored-out inside an internal function, it doesn't seem worth to factor it out even further..?

Reply:
@setrofim 
Please correct me if I am wrong but the same code snippet is repeated at
Line 433, Line 481, Line 635 and Line 667.

So I see four repetitions.

Reply:
Crap, sorry, I was looking at the switch inside  `deriveAlgoritm`. You are correct. My reply to the comment for line 479 applies to all of those --  the inner switch relies on the context provided by the outer switch, so factoring it out would neccesitate adding erro handling in the factored out function, _and_ everywhere that function is called, thus, IMO, is not worth it.

Reply:
We do not need to handle error in these calling part, as we have already verified using the outer switch.
The generic method, can return error for the case of unsupported algorithms.

Reply:
You are anyway ignoring the default error condition in the inner switch. So not much different.

This factored function is just extracting the mapped ES Curve from supported ES algorithms.
Hypothetically one can argue on this, but in practical terms this local function is only scoped to do little mapping to 
valid elements.

Reply:
> You are anyway ignoring the default error condition in the inner switch. So not much different.

I am not ignoring it. There is no error condition -- the `default` entry is not necessary  because the individual cases are exhaustive, given the context.  Factoring this out into a separate function removes the context, and necessitates a `default`, and therefore handling there of. The difference is that the function cannot make assumptions about its calling context, and the calling code _should not_ make the assumptions about the internal operation of the function.
 
> Hypothetically one can argue on this, but in practical terms [...]

Yes, brittleness is about hypotheticals -- it does not mean the code is incorrect right now, it means it is more likely to become incorrect in the future. Ignoring errors is a "bad smell", even if it  is not strictly speaking incorrect at this specific point in time.